### ISSUE TITLE
Key extraction

### ISSUE THREAD

**Issue by Max-Meldrum** [cda-group/arcon#75]
Rather than forcing Hash upon all ArconType's, look into having our own trait for extracting a key. That is, we don't only want to be able to hash keys, users may want their own custom logic for defining keys.

**Comment by Max-Meldrum** [comment]
```rust
// v1
trait ArconKey {
  get_key(&self) -> u64; // return 0 if no key?
}
// v2
trait ArconKey {
  get_key(&self) -> Option<u64>;
}

// if we enforce ArconKey on the ArconTypeBounds, then there are no issues.
// but, if we would like something like KeyBy<A: ArconType + ArconKey>, 
// then we would have to enforce it for all channel strategies with the way it is implemented currently.
pub enum ChannelStrategy<A>
where
    A: ArconType + ArconKey,
{
    Forward(forward::Forward<A>),
    KeyBy(key_by::KeyBy<A>), 
}
```

**Comment by Bathtor** [comment]
I think things should *always* have a key so I prefer v1. We can always fall back on a `derive(ArconKey)` based on a hash of the whole struct if no custom logic is needed, and if the strategy (like forward) doesn't need it, it will never be called, so it doesn't matter if it's unnecessarily expensive.

**Comment by mrobakowski** [comment]
I live v1 better. Either a type implements the trait and provides a sensible result or just doesn't. We don't really have to worry about the "no key" scenario because such types just shouldn't impl the trait.

... but I don't know how to do this while keeping `ChannelStrategy` an enum. If it was a trait, we could do something like:
```rust
struct Forward<A>;
impl<A: ArconType> ChannelStrategy<A> for Forward<A> {}

struct KeyBy<A>;
impl<A: ArconType + ArconKey> ChannelStrategy<A> for KeyBy<A> {}
```

Also, I don't really like the name `ArconKey`. `ArconKeyable` or just `HasKey` reflects what it does better.

**Comment by Max-Meldrum** [comment]
> I think things should _always_ have a key so I prefer v1. We can always fall back on a `derive(ArconKey)` based on a hash of the whole struct if no custom logic is needed, and if the strategy (like forward) doesn't need it, it will never be called, so it doesn't matter if it's unnecessarily expensive.

Was leaning towards this approach as well. But then there is no point in having a trait for it I guess?

```rust
pub trait ArconType: ArconTypeBounds
where
    Self: std::marker::Sized,
{
    /// Serialisation ID for Arcon's Unsafe In-flight serde
    const UNSAFE_SER_ID: SerId;
    /// Serialisation ID for Arcon's Reliable In-flight serde
    const RELIABLE_SER_ID: SerId;
    /// Current version of this ArconType
    const VERSION_ID: VersionId;

    fn get_key(&self) -> u64;
}
```

**Comment by mrobakowski** [comment]
I wonder if `prost` understands [custom options](https://developers.google.com/protocol-buffers/docs/proto#customoptions). Perhaps we could use that to specify the key field in the protobuf definition files.

**Comment by Bathtor** [comment]
> I wonder if `prost` understands [custom options](https://developers.google.com/protocol-buffers/docs/proto#customoptions). Perhaps we could use that to specify the key field in the protobuf definition files.

But what if it's a combination of fields that must be hashed to produce the actual key?

**Comment by mrobakowski** [comment]
If multiple ones are specified you may as well just plop them into a tuple and hash that. Anyway, that's a moot point because it seems like `prost` doesn't support custom options and reflection yet https://github.com/danburkert/prost/issues/256. 

Another idea is to look at the doc comments attached to the fields. They are preserved by `prost_build`, so `#[derive(ArconType)]` could just collect all the fields that have some marker, let's say `#arcon-key` in their docs and then generate an implementation of `fn get_key(&self)` based on those fields.

**Comment by Max-Meldrum** [comment]
It is already possible to hash on selected fields. However, it does not look that beautiful :)

```rust
#[derive(Arcon, prost::Message, Clone, abomonation_derive::Abomonation)]
#[arcon(unsafe_ser_id = 104, reliable_ser_id = 105, version = 1, keys = "id,price")]
pub struct Item {
  #[prost(uint64, tag = "1")]
   pub id: u64,
  #[prost(uint32, tag = "2")]
   pub price: u32,
}
```

**Comment by mrobakowski** [comment]
Yeah, when you write your datatype in Rust. I'm talking about the "generating from .proto3 files" part.

**Comment by Bathtor** [comment]
> It is already possible to hash on selected fields. However, it does not look that beautiful :)

Actually I think that's pretty good. Remember, we will be generating that code, so beautiful is not a requirement.

> Yeah, when you write your datatype in Rust. I'm talking about the "generating from .proto3 files" part.

Does this matter? We can always augment the generated Rust code with the correct hashing implementation, since this has to happen during our code generation anyway.

**Comment by Max-Meldrum** [comment]
Still applicable for generating from files. We will need some build file anyways I assume.

https://github.com/cda-group/arcon/blob/master/execution-plane/arcon/proto_tests/build.rs

**Comment by mrobakowski** [comment]
I guess it doesn't matter that much. But I think it would simplify the codegen if it went just:
`arc-script -> *.rs + *.proto3 -> binary`. What you describe would be more like
```
arc-script -----------> *.rs + *.proto3 ------> *.rs -> binary
        `-- whoops I gotta tweak those -------^
```
What I'm trying to say is that the protobuf definition files should have all the necessary information to generate the structs, including which fields should be considered when generating a key.

**Comment by Max-Meldrum** [comment]
sure, would be nice to be able to extract that directly.  But can't see how it would simplify things to a great degree for the codegen.  We will still need to add the custom arcon attributes per schema/build.

```rust
format!(
        "#[arcon(unsafe_ser_id = {}, reliable_ser_id = {}, version = {})]",
        unsafe_ser_id, reliable_ser_id, version
    )
```

**Comment by mrobakowski** [comment]
I could imagine a version of `#[derive(ArconType)]` that would get all of those attributes from doc comments on the struct generated by `prost`. Then that code could be the same for all schemas/builds. All the custom info would be hidden in the `.proto3` files. Less different things to generate by `arc-script`.

**Comment by Max-Meldrum** [comment]
```proto
myfile.proto

// unsafe_ver_id = 1, reliable_ser_id = 2, ...
message Hello {
  string msg = 1;
}
```

Do you mean something like this?

**Comment by mrobakowski** [comment]
Basically. `prost` will compile those into `/// doc comments` which are desugared by rustc into `#[doc = "doc comments"]` before feeding it into derive macros afaik. Then that can be processed by our derive macro.

**Comment by Max-Meldrum** [comment]
```rust
{ ident: Ident { ident: "doc", span: #0 bytes(0..0) }, arguments: None }] }, eq_token: Eq, lit: Str(LitStr { token: " hej = 10" }) })
```

I am able to get this out from:

```proto
// hej = 10
message SearchRequest {
  string query = 1;
  int32 page_number = 2;
  int32 result_per_page = 3;
}
```

**Comment by Max-Meldrum** [comment]
Resolved by #106
